### DISCUSSION TITLE
Populate invalid cast date

### DISCUSSION THREAD

**Discussion by karol71927** [mikro-orm/mikro-orm#5577] [General]
### Describe the bug

Hi, I found a Bug with populate. When I used find entity with populate records, records has date for example `2024-05-15T17:51:35.000Z` but when I find record without populate same record has date like this `2024-05-15T15:51:35.000Z`.
I am using timezone +02:00

### Reproduction

https://github.com/karol71927/mikro-orm-populate-issue

In example I am finding recipe with ingredients and then I am geting `Flour` date is `createdAt: 2024-05-15T17:51:35.000Z`. Then i wanto to get same ingredient and date is `createdAt: 2024-05-15T15:51:35.000Z,` 
In the first case I am using `populate` on recipe. In second case I am fetching ingredients with `find`

```ts    
    const em = this.orm.em.fork();

    const recipe = await em.findOne(
      Recipe,
      { title: 'My first recipe' },
      { populate: ['ingredients'] },
    );

    const withPopulate = recipe.ingredients
      .getItems()
      .find((x) => x.name === 'Flour');

    console.log(withPopulate);

    const ingredients = await em.find(Ingredient, {
      recipe: { id: recipe.id },
    });

    const withoutPopulate = ingredients.find((x) => x.name === 'Flour');
    console.log(withoutPopulate);
 ```

### What driver are you using?

@mikro-orm/mysql

### MikroORM version

6.2.6

### Node.js version

20

### Operating system

mac os

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by karol71927** [comment]
Hi,

To clarify the unexpected behaviour more

In MikroORM 5.x, with a configuration of TIMEZONE + 2 (although it can be any +/- hour), when I saved an entity with a date (stored as a datetime field in MySQL), for example, 2024-05-16T08:05:21.743Z, it was correctly saved in the database as 2024-05-16 10:05:21, which is expected.

Later, when I retrieved this entity using orm.em.findOne or a dedicated repository, the date on this entity remained as 2024-05-16T08:05:21.743Z, as expected.

This behavior persists in MikroORM 6.x.

However, the unexpected behavior occurs when retrieving an entity with its children. In 5.x, if the children have dates associated with them, the behavior mirrors that of the parent entity. That is, when saving the main entity with children and associated dates, such as 2024-05-16T08:05:21.743Z, the dates on the children were correctly saved in the database as 2024-05-16 10:05:21. Upon retrieval via populate, the dates on the children matched the timezone-adjusted time of the main entity, appearing as 2024-05-16T08:05:21.743Z.

However, in MikroORM 6.x, performing the same actions results in incorrectly calculated or casted dates on the children. Even though the main entity's date is properly saved in the database as 2024-05-16 10:05:21, when using findOne on the main entity and populating its children, the dates on the children are displayed as 2024-05-16T10:05:21.743Z. This discrepancy means that the child entity's time does not match the timezone-adjusted time of the main entity, resulting in a difference of additional two hours (2024-05-16 12:05:21 instead of 2024-05-16 10:05:21).

**Reply by B4nan** [reply]
Can you simplify the repro? It should be a simple script, one file, no webservers needed. I want to just run it and see what you see.

https://github.com/mikro-orm/reproduction

**Comment by B4nan** [comment]
Tried your repro again and it seems to be working fine on my end, all the dates are same. 

Here is a simplified version, which also works fine:

```ts
import { Collection, Entity, ManyToOne, MikroORM, OneToMany, PrimaryKey, Property } from '@mikro-orm/mysql';

@Entity()
class Recipe {

  @PrimaryKey()
  id!: number;

  @Property()
  title: string;

  @Property({ onCreate: () => new Date() })
  createdAt!: Date;

  @OneToMany(() => Ingredient, ingredient => ingredient.recipe)
  ingredients = new Collection<Ingredient>(this);

  constructor(title: string) {
    this.title = title;
  }

}

@Entity()
class Ingredient {

  @PrimaryKey()
  id!: number;

  @Property()
  name: string;

  @Property()
  quantity: number;

  @Property({
    onCreate: () => new Date(),
  })
  createdAt!: Date;

  @ManyToOne(() => Recipe)
  recipe!: Recipe;

  constructor(name: string, quantity: number) {
    this.name = name;
    this.quantity = quantity;
  }

}

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    entities: [Recipe],
    dbName: '5577',
    port: 3308,
    timezone: '+02:00',
    debug: true,
  });
  await orm.schema.refreshDatabase();
});

afterAll(() => orm.close(true));

test('5577', async () => {
  const r = new Recipe('My first recipe');
  r.ingredients.add(new Ingredient('Flour', 500));
  r.ingredients.add(new Ingredient('Sugar', 200));
  r.ingredients.add(new Ingredient('Milk', 200));
  orm.em.persist(r);
  await orm.em.flush();
  orm.em.clear();

  const recipe = await orm.em.findOneOrFail(
    Recipe,
    { title: 'My first recipe' },
    { populate: ['ingredients'] },
  );

  const withPopulate = recipe.ingredients
    .getItems()
    .find(x => x.name === 'Flour');

  console.log(withPopulate);

  const ingredients = await orm.em.find(Ingredient, {
    recipe: { id: recipe.id },
  });

  const withoutPopulate = ingredients.find(x => x.name === 'Flour');
  console.log(withoutPopulate);
});
```

**Reply by karol71927** [reply]
I found an issue with reproduction, dates are the same because there is working identity map. When I disabled identity map reproduction is working. Here is repository with reproduction https://github.com/karol71927/reproduction-mikro-orm. You can run example with `docker compose up --exit-code-from app`. For me test result looks like this 
![image](https://github.com/mikro-orm/mikro-orm/assets/58816240/5f624950-31d2-428b-b4fa-bc463f4adcd6)

**Reply by B4nan** [reply]
The reproduction explicitly calls `em.clear()`, so nope, its not about that.

<img width="444" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/615580/e8415830-bb3d-411f-8fb0-45b7f3dfa4fa">

**Reply by karol71927** [reply]
did you updated test case to code from reproduction?
```ts
test("5577", async () => {
  const r = new Recipe("My first recipe");
  r.ingredients.add(new Ingredient("Flour", 500));
  r.ingredients.add(new Ingredient("Sugar", 200));
  r.ingredients.add(new Ingredient("Milk", 200));
  orm.em.persist(r);
  await orm.em.flush();
  orm.em.clear();

  const recipe = await orm.em.findOneOrFail(
    Recipe,
    { title: "My first recipe" },
    { populate: ["ingredients"], disableIdentityMap: true }
  );

  const withPopulate = recipe.ingredients
    .getItems()
    .find((x) => x.name === "Flour");

  console.log(withPopulate);

  const ingredients = await orm.em.find(
    Ingredient,
    {
      recipe: { id: recipe.id },
    },
    { disableIdentityMap: true }
  );

  const withoutPopulate = ingredients.find((x) => x.name === "Flour");
  console.log(withoutPopulate);

  console.log(withPopulate?.createdAt, withoutPopulate?.createdAt);

  expect(withPopulate?.createdAt).toEqual(withoutPopulate?.createdAt);
});
```

**Reply by B4nan** [reply]
Yes, this is from your version.

**Reply by karol71927** [reply]
When I’m running it locally it works fine on my computer (I have timezone +02:00, Europe/Warsaw).
But when I am running it in docker, entities dates with and without populate are still different . The test fail.
In docker container (The one from repro I created) I have UTC+0 timezone. Can you check what happens if you run it in docker?
Result from docker container
![image](https://github.com/mikro-orm/mikro-orm/assets/58816240/70a73489-8abe-4084-8e59-e4293c6a73d6)
Result from local run 
![image](https://github.com/mikro-orm/mikro-orm/assets/58816240/3858cac1-9e7f-4359-ad1f-177ba9502a65)

**Reply by B4nan** [reply]
I would guess the docker container doesn't have the timezone as in your host system.

**Reply by B4nan** [reply]
My MySQL instance is running in docker already btw, I am using the one from the ORM test suite.

https://github.com/mikro-orm/mikro-orm/blob/master/docker-compose.yml#L10

**Reply by karol71927** [reply]
Yeah in both cases MySql is inside docker container. The difference is that test app is in container

**Reply by B4nan** [reply]
Can you try using `select-in` loading strategy? If it works that way but the (default) joined strategy fails, I think I know where to look.

**Reply by karol71927** [reply]
it passed with 
```ts
 const recipe = await orm.em.findOneOrFail(
    Recipe,
    { title: "My first recipe" },
    {
      populate: ["ingredients"],
      disableIdentityMap: true,
      strategy: LoadStrategy.SELECT_IN,
    }
  );
  ```

**Reply by karol71927** [reply]
and JOINED strategy fails

**Reply by B4nan** [reply]
Yeah, the root entities are mapped this way which respects the timezone option:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/EntityComparator.ts#L288

But the joined relations are handled dynamically elsewhere:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/AbstractSqlDriver.ts#L372

And there we ignore the explicit timezone.

**Reply by B4nan** [reply]
And now I understand why the repro is passing, I am also in UTC+2. Setting the `timezone` option to another value makes it fail.

**Reply by B4nan** [reply]
How do you plan to deal with DST? If the absolute timezone shift changes you would need to change the ORM config option and reload the app for it to have an effect, that doesn't sound very maintainable. I guess that's why most people just enforce UTC on the server and deal with timezones when presenting the data instead.

**Reply by B4nan** [reply]
This should do it: 7f9bb0be291088b0098b839753bc7ca65fed1e99

**Reply by karol71927** [reply]
For me using timezone +2 is fine. I'm not changing it. It's only for backend but frontend never shows dates. Enforcing UTC seems to be a good idea. Thank you for fix 😄
